### PR TITLE
Synchronize Bluetooth documentation between development-publish and publish branch

### PR DESCRIPTION
**Pull request by geza-pycom**
To be merged when the 2 features from the beginning of the list below are released in a stable release.

In this PR the currently latest version of Bluetooth documentation from "development-publish" is synchronized with "publish" branch as there were lots of not documented, but already released, changes:
- Documentation of feature "BLE characteristic update messages are lost if they sent too frequently", released as beta in: v1.20.2.rc2, publish-development PR: https://github.com/pycom/pycom-documentation/pull/215
- Documentation of feature "Implemented configuration options for BLE security", not released yet in any development version, publish-development PR: https://github.com/pycom/pycom-documentation/pull/225
- Documentation of the following, already officially released in 1.20.1, changes:
    - Adding "modem_sleep" parameter to Bluetooth.init()
    - Adding "timeout" parameter to Bluetooth.connect()
    - Adding documentation of Bluetooth.set_advertisement_params() new API
    - Adding documentation of Bluetooth.tx_power() new API

### CODE REVIEW

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 1]
```diff
@@ -60,14 +61,18 @@ bluetooth = Bluetooth()
...
+#### bluetooth.init(id=0, mode=Bluetooth.BLE, antenna=None, modem\_sleep=True, pin=None, privacy=True, secure\_connections=True, mtu=200)
 
-* `id` Only one Bluetooth peripheral available so must always be 0
-* `mode` currently the only supported mode is `Bluetooth.BLE`
-* `antenna` selects between the internal and the external antenna. Can be either`Bluetooth.INT_ANT`, `Bluetooth.EXT_ANT`.
-* `secure` enables or disables the GATT Server security features.
+* `id` Only one Bluetooth peripheral available so must always be 0.
+* `mode` currently the only supported mode is `Bluetooth.BLE`.
+* `antenna` selects between the internal and the external antenna. Can be either `Bluetooth.INT_ANT`, `Bluetooth.EXT_ANT`.
+* `modem_sleep` Enables or Disables BLE modem sleep, Disable modem sleep as a workaround when having Crashes due to flash cache being disabled, as this prevents BLE task saving data in external RAM while accesing external flash for R/W.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:69]: 1st: Enables or Disables BLE modem sleep. Disable modem sleep is a workaround when experiencing crashes due to the flash cache being disabled. It prevents the BLE task saving data in the external RAM while accessing the external flash for R/W.

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 2]
```diff
@@ -60,14 +61,18 @@ bluetooth = Bluetooth()
...
+* `pin` a one to six digit number (`0`-`9`) to connect to the GATT Sever. Setting any valid pin, GATT Server security features are activated.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:70]: 2nd: By setting a valid pin, the GATT Server security features are activated.

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 3]
```diff
@@ -60,14 +61,18 @@ bluetooth = Bluetooth()
...
+* `privacy` Enables or Disables local privacy settings so address will be random or public.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:71]: 3rd: Enables or disables the local privacy settings so the address will be random or public.

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 4]
```diff
@@ -60,14 +61,18 @@ bluetooth = Bluetooth()
...
+* `secure_connections` Enables or Disables Secure Connections and MITM Protection.
+* `mtu` Maximum Transmission Unit (MTU) is the maximum length of an ATT packet. Value must be between `23` and `200`.
 
-  With our development boards it defaults to using the internal antenna, but in the case of an OEM module, the antenna pin (`P12`) is not used, so it's free to be used for other things.
+With our development boards it defaults to using the internal antenna, but in the case of an OEM module, the antenna pin (`P12`) is not used, so it's free to be used for other things.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:75]: 4th: Our development boards' default is to use the internal antenna.  However, with the OEM module, the antenna pin (`P12`) is not used so it's free to be used for other things.

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 5]
```diff
@@ -84,7 +89,7 @@ Pin('P12', mode=Pin.OUT)(True)
 
 Disables the Bluetooth radio.
 
-### bluetooth.start\_scan(timeout)
+#### bluetooth.start\_scan(timeout)
 
 Starts performing a scan listening for BLE devices sending advertisements. This function always returns immediately, the scanning will be performed on the background. The return value is `None`. After starting the scan the function `get_adv()` can be used to retrieve the advertisements messages from the FIFO. The internal FIFO has space to cache 16 advertisements.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:94]: 5th: Starts performing a scan, listening for BLE devices sending advertisements. This function always returns immediately and the scanning will be performed in the background. The return value is `None`. 
After starting the scan, the function `get_adv()` can be used to retrieve the advertisement's messages from the FIFO. The internal FIFO has space to cache 16 advertisements.

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 6]
```diff
@@ -95,20 +100,19 @@ The arguments are:
...
 
-### bluetooth.isscanning()
+#### bluetooth.isscanning()
 
 Returns `True` if a Bluetooth scan is in progress. `False` otherwise.
 
-### bluetooth.get\_adv()
+#### bluetooth.get\_adv()
 
 Gets an named tuple with the advertisement data received during the scanning. The tuple has the following structure: `(mac, addr_type, adv_type, rssi, data)`
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:117]: 6th: Gets a named tuple with the advertisement data received during the scan. The tuple has the following structure:

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 7]
```diff
@@ -131,11 +134,11 @@ adv = bluetooth.get_adv() #
 ubinascii.hexlify(adv.mac) # convert hexadecimal to ascii
 ```
 
-### bluetooth.get\_advertisements()
+#### bluetooth.get\_advertisements()
 
 Same as the `get_adv()` method, but this one returns a list with all the advertisements received.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:139]: 7th: Same as the `get_adv()` method, but instead returns a list of all the advertisements received.

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 8]
```diff
@@ -131,11 +134,11 @@ adv = bluetooth.get_adv() #
...
 
-### bluetooth.resolve\_adv\_data(data, data\_type)
+#### bluetooth.resolve\_adv\_data(data, data\_type)
 
 Parses the advertisement data and returns the requested `data_type` if present. If the data type is not present, the function returns `None`.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:143]: 8th: Passes the advertisement data and returns the requested `data_type` if present. If the data type is not present, the function returns `None`.

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 9]
```diff
@@ -166,18 +168,20 @@ while bluetooth.isscanning():
             print(ubinascii.hexlify(mfg_data))
 ```
 
-### bluetooth.connect(mac\_addr)
+#### bluetooth.set\_pin()
 
-Opens a BLE connection with the device specified by the `mac_addr` argument. This function blocks until the connection succeeds or fails. If the connections succeeds it returns a object of type `GATTCConnection`.
+Configures a new PIN to be used by the device. The PIN is a 1-6 digit length decimal number, if less than 6 digits are given the missing leading digits are considered as 0. E.g. 1234 becomes 001234. When a new PIN is configured, the information of all previously bonded device is removed and the current connection is terminated. To restart advertisement the advertise() must be called after PIN is changed.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:173]: 9th: Configures a new PIN to be used by the device. The PIN is a one to six digit decimal number. If less than 6 digits are given, the missing leading digits are considered as 0 (for example, 1234 becomes 001234.) When a new PIN is configured, all previously bonded device information is removed and the current connection is terminated. To restart the advertisement,  the advertise() must be called after the PIN is changed.

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 10]
```diff
@@ -166,18 +168,20 @@ while bluetooth.isscanning():
...
 
-Connections are initiated by the central device. There is a maximum of 4 simultaneous connections.
+#### bluetooth.connect(mac\_addr, timeout=None)
 
-```python
+* `mac_addr` is the address of the remote device to connect
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:177]: 10th: is the address of the remote device that's connected

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 11]
```diff
@@ -166,18 +168,20 @@ while bluetooth.isscanning():
...
+* `timeout` specifies the amount of time in milliseconds to wait for the connection process to finish. If not given then no timeout is applied The function blocks until the connection succeeds or fails (raises OSError) or the given `timeout` expires (raises `Bluetooth.timeout TimeoutError`). If the connections succeeds it returns a object of type `GATTCConnection`.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:178]: 11th: specifies the amount of time in milliseconds to wait for the connection process to finish. If not given, then no timeout is applied. The function blocks until the connection succeeds or fails (raises OSError) or the given `timeout` expires (raises `Bluetooth.timeout TimeoutError`). If the connections succeeds, it returns an object of type `GATTCConnection`.

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 12]
```diff
@@ -166,18 +168,20 @@ while bluetooth.isscanning():
...
 
+```python
 bluetooth.connect('112233eeddff') # mac address is accepted as a string
 ```
 
-### bluetooth.callback(trigger=None, handler=None, arg=None)
+#### bluetooth.callback(trigger=None, handler=None, arg=None)
 
 Creates a callback that will be executed when any of the triggers occurs. The arguments are:
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:186]: 12th: Creates a callback that will be executed when any of the triggers occur. The arguments are:

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 13]
```diff
@@ -212,7 +215,7 @@ bluetooth.callback(trigger=Bluetooth.CLIENT_CONNECTED | Bluetooth.CLIENT_DISCONN
 bluetooth.advertise(True)
 ```
 
-### bluetooth.set\_advertisement(\* , name=None, manufacturer\_data=None, service\_data=None, service\_uuid=None)
+#### bluetooth.set\_advertisement(\* , name=None, manufacturer\_data=None, service\_data=None, service\_uuid=None)
 
 Configure the data to be sent while advertising. If left with the default of `None` the data won't be part of the advertisement message.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:220]: 13th: Configures the data to be sent whilst advertising. If left with the default of `None`,  the data won't be part of the advertisement message.

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 14]
```diff
@@ -226,15 +229,27 @@ The arguments are:
...
+#### bluetooth.set\_advertisement\_params(\* , adv\_int\_min=0x20, adv\_int\_max=0x40, adv\_type=Bluetooth.ADV\_TYPE\_IND, own\_addr\_type=Bluetooth.BLE\_ADDR\_TYPE\_PUBLIC, channel\_map=Bluetooth.ADV\_CHNL\_ALL, adv\_filter\_policy=Bluetooth.ADV\_FILTER\_ALLOW\_SCAN\_ANY\_CON\_ANY)
+
+Configure the parameters used when advertising.
+
+The arguments are:
+
+* `adv_int_min` is the minimum advertising interval for undirected and low duty cycle directed advertising.
+* `adv_int_max` is the maximum advertising interval for undirected and low duty cycle directed advertising.
+* `adv_type` is the advertising type.
+* `own_addr_type` is the owner bluetooth device address type.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:244]: 14th: is the owner's bluetooth device address type.

File: content/firmwareapi/pycom/network/bluetooth/_index.md [hunk 15]
```diff
@@ -246,20 +261,35 @@ The arguments are:
...
 
-### bluetooth.disconnect\_client()
+#### bluetooth.disconnect\_client()
 
 Closes the BLE connection with the client.
 
+### bluetooth.tx\_power(type, level)
+
+Gets or sets the TX Power level.
+If called with only `type` parameter it returns with the current value belonging to the given type.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/_index.md:274]: 15th: If called with only `type` parameter,  it returns with the current value belonging to the given type.

File: content/firmwareapi/pycom/network/bluetooth/gattscharacteristic.md [hunk 16]
```diff
@@ -21,70 +21,63 @@ characteristic.value(123) # set characteristic value to an integer with the valu
...
 * `handler` is the function that will be executed when the callback is triggered.
 * `arg` is the argument that gets passed to the callback. If nothing is given, the characteristic object that owns the callback will be used.
 
-An example of how this could be implemented can be seen in the [`characteristic.events()` ](gattscharacteristic.md#characteristic-events)section.
+Beyond the `arg` a tuple (called `data`) is also passed to `handler`. The tuple consists of (event, value), where `event` is the triggering event and `value` is the value strictly belonging to the `event` in case of a WRITE event. If the `event` is not a WRITE event, the `value` has no meaning.
 
-#### characteristic.events()
-
-Returns a value with bit flags identifying the events that have occurred since the last call. Calling this function clears the events.
+We recommend getting both the `event` and new `value` of the characteristic via this tuple, and not via `characteristic.event()` and `characteristic.value()` calls in the context of the `handler` to make sure no event and value is lost.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/gattscharacteristic.md:38]: 16th: We recommend getting both the `event` and new `value` of the characteristic via this tuple, and not via `characteristic.event()` and `characteristic.value()` calls in the context of the `handler`. This is to make sure that no event and value is lost.

File: content/firmwareapi/pycom/network/bluetooth/gattsservice.md [hunk 17]
```diff
@@ -27,12 +27,11 @@ Stops the service if previously started.
 Creates a new characteristic on the service. Returns an object of the class `GATTSCharacteristic`. The arguments are:
 
 * `uuid` is the UUID of the service. Can take an integer or a 16 byte long string or bytes object.
-* `permissions` configures the permissions of the characteristic. Takes an integer with a combination of the flags.
+* `permissions` configures the permissions of the characteristic. Takes an integer with a combination of the flags. When bluetooth object is initialized with PIN, read and write permissions are set to encrypted. Setting PIN later with set_pin() call does not affect the permissions of the already existing characteristics, thus they will remain not secured.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/bluetooth/gattsservice.md:30]: 17th: When the bluetooth object is initialized with the PIN, read and write permissions are set to encrypted. Setting the PIN later with the set_pin() call does not affect the permissions of the already existing characteristics, thus they will remain unsecured.

**Review by lilycey — COMMENTED** [review]
Minor changes for readability

**Comment by geza-pycom** [comment]
@lilycey : let me not modify these things as this is just a synchronization commit between the documentation of the development release and stable release. These corrections should be made in both documentation.

**Review by peter-pycom — APPROVED** [review]

**Pull request merged**